evm/p2p/discovery.py

```diff
@@ -0,0 +1,320 @@
+
+    def stop(self):
+        self.logger.info('stopping discovery')
+        self.transport.close()
+
+    def receive(self, address, message):
+        try:
+            remote_pubkey, cmd_id, payload, mdc = _unpack(message)
+            # Note: as of discovery version 4, expiration is the last element for all
+            # packets. This might not be the case for a later version, but just popping
+            # the last element is good enough for now.
+            expiration = rlp.sedes.big_endian_int.deserialize(payload.pop())
+            if time.time() > expiration:
+                raise PacketExpired()
+        except DefectiveMessage as e:
+            self.logger.error('error unpacking message: {}'.format(e))
+            return
+
+        cmd = CMD_ID_MAP[cmd_id]
+        if len(payload) != cmd.elem_count - 1:
+            self.logger.error('invalid {} payload: {}'.format(cmd.name, payload))
+            return
+        node = kademlia.Node(remote_pubkey, address)
+        handler = self._get_handler(cmd)
+        handler(node, payload, mdc)
+
+    def recv_pong(self, node, payload, mdc):
+        echoed = payload[1]
+        self.kademlia.recv_pong(node, echoed)
+
+    def recv_neighbours(self, node, payload, mdc):
+        self.kademlia.recv_neighbours(node, _extract_nodes_from_payload(payload[0]))
+
+    def recv_ping(self, node, payload, mdc):
+        self.kademlia.recv_ping(node, mdc)
+
+    def recv_find_node(self, node, payload, mdc):
+        self.logger.debug('<<< find_node from {}'.format(node))
+        target = big_endian_to_int(payload[0])
+        self.kademlia.recv_find_node(node, target)
+
+    def send_ping(self, node):
+        self.logger.debug('>>> pinging {}'.format(node))
+        version = rlp.sedes.big_endian_int.serialize(PROTO_VERSION)
+        payload = [version, self.address.to_endpoint(), node.address.to_endpoint()]
+        message = _pack(CMD_PING.id, payload, self.privkey)
+        self.send(node, message)
+        return message[:32]  # return the MDC to identify pongs
+
+    def send_find_node(self, node, target_node_id):
+        target_node_id = int_to_big_endian(
+            target_node_id).rjust(kademlia.k_pubkey_size // 8, b'\0')
+        self.logger.debug('>>> find_node to {}'.format(node))
+        message = _pack(CMD_FIND_NODE.id, [target_node_id], self.privkey)
+        self.send(node, message)
+
+    def send_pong(self, node, token):
+        self.logger.debug('>>> ponging {}'.format(node))
+        payload = [node.address.to_endpoint(), token]
+        message = _pack(CMD_PONG.id, payload, self.privkey)
+        self.send(node, message)
+
+    def send_neighbours(self, node, neighbours):
+        nodes = []
+        neighbours = sorted(neighbours)
+        for n in neighbours:
+            l = n.address.to_endpoint() + [n.pubkey]
+            nodes.append(l)
+
+        max_neighbours = self._get_max_neighbours_per_packet()
+        if len(nodes) <= max_neighbours:
```

Comment:
Other than the minor optimization, I think this code will be noticeably easier to read if we keep this logic in a single loop (i.e. removing this part and only keeping the part in `else` statement.

Reply:
Indeed, good catch!